src/System.Management.Automation/engine/runtime/Operations/MiscOps.cs

```diff
@@ -1901,10 +1949,13 @@ internal static bool ReportErrorRecord(IScriptExtent extent, RuntimeException rt
 
             context.ShellFunctionErrorOutputPipe.Add(errorWrap);
 
+            // set the value of $? here in case it is reset in error reporting.
+            context.QuestionMarkVariableValue = false;
```

Comment:
I have no idea. That logic pre-dates these changes. I just carried it forward.

Reply:
This line was newly added to this method, and that means you think this method could change `context.QuestionMarkVariableValue`, right? I would like to understand the necessity of this change.

Reply:
Ok, I see the source of confusion now. That line (and the comment that precedes it) was moved into that method from the `CheckActionPreference` method that invokes it. [Here is the same line before my changes](https://github.com/PowerShell/PowerShell/blob/bd6fdae73520931f0d27a29d6290e18761772141/src/System.Management.Automation/engine/runtime/Operations/MiscOps.cs#L1642-L1643). With the other places where `QuestionMarkVariableValue` needs to be set now properly setting it, and given the comment before it, it made much more sense for that line to be inside of `ReportErrorRecord` (assuming it is actually necessary). I didn't want to discard the line, just tidy how it is used up a little, and make sure that `$?` gets set whenever `ReportErrorRecord` is invoked in case there is risk of it being reset in error reporting.